### DISCUSSION TITLE
is there a flowchart / diagram of how freetube works?

### DISCUSSION THREAD

**Discussion by ceeb9** [FreeTubeApp/FreeTube#2779] [Q&A] [answered]
i'd like to contribute, but without combing through the code for hours i don't feel like i would get a good understanding of how each piece fits together. can someone explain the program flow? 
for example, on open, function x makes a request for y, then stores the reply in z, then it's processed by the next function, etc. etc.
my experience with web dev specifically is limited, but i'm still eager to help out.

**Comment by PikachuEXE** [comment]
Maybe you can pick one specific function you want to study/update so we can point you to code locations?
Easier to start somewhere (and probably small/simple)

**Comment by absidue** [comment, marked as answer]
Hi @ceeb9 are there any features you would like to implement or bugs you want to fix? We'll happily point you in the right direction.

FreeTube uses the Vue 2 framework, which is designed around views (pages) and components. Each view can contain HTML and various components and each component can contain just HTML or other components as well. In FreeTube the views are in `src/renderer/views` and the components in `src/renderer/components`.

Examples of views are the subscriptions page or the watch page
Examples of components are the video player, the video description, the comments, etc

**Reply by ceeb9** [reply]
ah that's a lot simpler than i thought, thanks for the pointer.

what i was thinking of doing was implementing some kind of system to generate recommendations, using the tagging info that youtube provides. (the rounded bubbles on the home page allow you to select from different categories, this is the tagging data i mean.) based on that, you could check the most popular/trending pages for videos with tags that overlap with those on the subscriptions page, and get some, albeit basic, recommendations.

the two major questions i have are
1. how does freetube store this kind of data? (history, subscribed channels etc.)
2. does youtube even expose data like this? i assume it is stored _somewhere_, given that it is in plaintext on the homepage, and not hidden away in the synapses of the machine learning recommendations algorithm.

if you could help me with those two that would be great.

**Reply by ChunkyProgrammer** [reply]
1) we store this data in a db using a node module called `nedb` (well a fork of nedb)
List of channels a user is subscribed to:
`src/renderer/store/modules/profiles.js` 
(a subscription is stored with a name, id and a thumbnail)
List of videos from the channels a user is subscribed to
`src/renderer/store/modules/subscriptions.js`

2) youtube does expose tags for channels and videos we currently use ytdl-core for videos and yt-channel-info for channels (we plan on migrating to a new library). In yt-channel-info, it'd returns an object with a tags attribute in it: https://github.com/FreeTubeApp/yt-channel-info/blob/e5d00604c80eba0f46194df664e4a51271d100b7/app/youtube-grabber.js#L157
I am not as familiar with ytdl-core but it should also be possible.

My thoughts:
- i don't think using the chips/rounded bubbles on the YouTube home page would be possible because FreeTube doesnt use YouTube accounts so YouTube won't know your subscription list
- I think the most viable option would be to do something similar to NewPipe for the tags, which is to display the tags in the video info + the channel about page and to search for the tag when you click on one.

![](https://user-images.githubusercontent.com/78101139/210287735-11b4c5b1-1f6a-4687-a80f-466a2d059f29.png)

@absidue is there anything else you'd want to add or do you think there's anything I missed?

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
@ceeb9 do you think that your questions have been answered?
